galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,526 @@
+        cos_dec1 = np.cos(delta1.rad())
+        sin_dec2 = np.sin(delta2.rad())
+        cos_dec2 = np.cos(delta2.rad())
+        cos_dalpha1 = np.cos((alpha1 - use_ra).rad())
+        sin_dalpha1 = np.sin((alpha1 - use_ra).rad())
+        cos_dalpha2 = cos_dalpha1
+        sin_dalpha2 = sin_dalpha1
+
+        # Can compute sin_theta1 and sin_theta2 directly without first computing theta1:
+        sin_theta1 = sin_dec1*sin_decp + cos_dec1*cos_decp*cos_dalpha1
+        sin_theta2 = sin_dec2*sin_decp + cos_dec2*cos_decp*cos_dalpha2
+        # These should always be in range 0-1, but look out for roundoff error:
+        if sin_theta1 < 0.: sin_theta1 = 0.
+        if sin_theta1 > 1.: sin_theta1 = 1.
+        if sin_theta2 < 0.: sin_theta2 = 0.
+        if sin_theta2 > 1.: sin_theta2 = 1.
+        # This is always in the range 0-90, so can compute cos_theta simply:
+        cos_theta1 = np.sqrt(1.-sin_theta1*sin_theta1)
+        cos_theta2 = np.sqrt(1.-sin_theta2*sin_theta2)
+
+        phi1 = phi_p + \
+            np.arctan2(-cos_dec1*sin_dalpha1, 
+                        sin_dec1*cos_decp-cos_dec1*sin_decp*cos_dalpha1)*galsim.radians
+        phi2 = phi_p + \
+            np.arctan2(-cos_dec2*sin_dalpha2,
+                        sin_dec2*cos_decp-cos_dec2*sin_decp*cos_dalpha2)*galsim.radians
+
+        # Convert to X-Y positions:
+        # Don't need to protect against divide by zero as theta is always near 90 degrees.
+        xtp1 = cos_theta1*np.sin(phi1.rad())*galsim.radians/sin_theta1
+        ytp1 = -cos_theta1*np.cos(phi1.rad())*galsim.radians/sin_theta1
+        xtp2 = cos_theta2*np.sin(phi2.rad())*galsim.radians/sin_theta2
+        ytp2 = -cos_theta2*np.cos(phi2.rad())*galsim.radians/sin_theta2
+        dxtp = xtp2 - xtp1
+        dytp = ytp2 - ytp1
+
+        # Finally have the ingredients for computing the position angle of this SCA Y axis.
+        pa_sca = np.arctan2(-cos_sca_rot*dxtp.rad()+sin_sca_rot*dytp.rad(),
+                             sin_sca_rot*dxtp.rad()+cos_sca_rot*dytp.rad())*galsim.radians
+        #print 'Orig calculation: ',pa_sca.rad()
+
+        # Compute CD coefficients: extract the linear terms from the a_sip, b_sip arrays.  These
+        # linear terms are stored in the SIP arrays for convenience, but are defined differently.
+        # The other terms have been divided by the linear terms, so that these become pure
+        # multiplicative factors. There is no need to change signs of the SIP coefficents associated
+        # with odd powers of X! Change sign of a10, b10 because the tangent-plane X pixel coordinate
+        # has sign opposite to the detector pixel X coordinate, and this transformation maps pixels
+        # to tangent plane.
+        a10 = -a_sip[i_sca,1,0]
+        a11 = a_sip[i_sca,0,1]
+        b10 = -b_sip[i_sca,1,0]
+        b11 = b_sip[i_sca,0,1]
+
+        # Rotate by pa_fpa.
+        cos_pa = np.cos(pa_sca.rad())
+        sin_pa = np.sin(pa_sca.rad())
```

Comment:
Ouch.  I see. It didn�t matter before because the code wasn�t using those variables in the calculations for each SCA, but now they are, so they got overwritten at the end of the calculations for the 1st SCA.  Thanks!